File: 1635-number-of-good-pairs/number-of-good-pairs.py
# Brute Force Code & Optimal Code
class Solution:
    def numIdenticalPairs(self, nums: List[int]) -> int:
        # dictionary to store the frequency of each number in the array
        mp = {}
        # count how many times each number appears
        for num in nums:
            # if num is already present increases its frequncy by 1 otherwise start its frequeny at 1.
            mp[num] = mp.get(num, 0) + 1
        # store the total number of good pairs
        result = 0
        # check every distinct numbers
        for num in mp:
            # get the number of times this values appears in the array
            count = mp[num]
            # if number appears count times choose any 2 occurence to form pair 
            result += (count * (count - 1) // 2)
        # return the total number of pairs where both indices contain the same values
        return result 

# Time Complexity : O(N)
# Space Complexity : O(1)

File: 1635-number-of-good-pairs/number-of-good-pairs.cpp
// Brute Force Code & Optimal Code
class Solution {
public:
    int numIdenticalPairs(vector<int>& nums) {
        // dictionary to store the frequency of each number in the array
        unordered_map<int, int> mp;
        // count how many times each number appears
        for (int num : nums) {
            // if num is already present increase its frequency by 1 otherwise start its frequency at 1
            mp[num]++;
        }
        // store the total number of good pairs
        int result = 0;
        // check every distinct number
        for (auto& [num, count] : mp) {
            // if a number appears count times choose any 2 occurrences to form a pair
            result += (count * (count - 1)) / 2;
        }
        // return the total number of pairs where both indices contain the same value
        return result;
    }
};

// Time Complexity : O(N)
// Space Comlexity : O(1)